src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -652,15 +660,130 @@ int PeerReplayer::remote_mkdir(const std::string &epath, const struct ceph_statx
   return 0;
 }
 
+PeerReplayer::FileTransferHandlerThreadPool::FileTransferHandlerThreadPool(
+    int _num_threads, int _max_concurrent_large_file,
+    uint64_t _file_size_threshold)
+    : num_threads(_num_threads), max_concurrent_large_file(std::min(
+                                     _num_threads, _max_concurrent_large_file)),
+      file_size_threshold(_file_size_threshold), workers(_num_threads),
+      stop_flag(true) {
+  max_token_sum = file_size_threshold * num_threads * max_concurrent_large_file;
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::activate() {
+  ceph_assert(stop_flag.load());
+  stop_flag = false;
+  for (auto &worker : workers) {
+    worker = std::thread(&FileTransferHandlerThreadPool::run, this);
+  }
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::deactivate() {
+  ceph_assert(!stop_flag.load());
+  stop_flag = true;
+  pick_task.notify_all();
+  give_task.notify_all();
+  exec_task.notify_all();
+  for (auto &worker : workers) {
+    if (worker.joinable()) {
+      worker.join();
+    }
+  }
+  drain_queue();
```

Comment:
Problem is not about the `current_token_sum`. I maintained a atomic counter `rem_file_transfer` for each directory locally and send reference of that counter to the file transfer threads. My goal is to even all the file transfer's are posted to the queue, `do_synchronize` function should wait until all the file associated with this directory are executed.
```
counter++;
    task_queue.emplace(
        [f = std::forward<F>(f), ... args = std::forward<Args>(args), &counter,
         &dir_cv](bool dont_execute) mutable {
          if (!dont_execute) {
            f(std::move(args)...);
          }
          --counter;
          if (counter == 0) {
            dir_cv.notify_one();
          }
        },
        token);
```

Here is the counter. But when a stop_flag is raised, it doesn't know about the counter, so I though best way is to just drain the queue by calling the function f(dont_execute = true). It will don't execute the transfer just decrease the counter and let `do_synchronize` exit safely.